File: GoLang/剑指Offer/easy/剑指Offer_54_二叉搜索树的第k大节点.go
package easy

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func kthLargest(root *TreeNode, k int) int {
	var ans int
	helper(root, &k, &ans)
	return ans
}

func helper(root *TreeNode, k *int, ans *int) {
	if root == nil || *k < 1 {
		return
	}

	helper(root.Right, k, ans)
	if *k == 1 {
		*ans = root.Val
	}
	*k--
	helper(root.Left, k, ans)
}

File: Java/剑指Offer/easy/剑指Offer_54_二叉搜索树的第k大节点.java
package easy;

/**
 * Created by lwd at 2021/8/6
 *
 * @Description:
 */
public class 剑指Offer_54_二叉搜索树的第k大节点 {
    private int index;
    private int ans;

    public int kthLargest(TreeNode root, int k) {
        index = k;
        preOrder(root);
        return ans;
    }

    private void preOrder(TreeNode root) {
        if (root == null || index < 0) {
            return;
        }

        preOrder(root.right);
        index--;
        if (index == 0) {
            ans = root.val;
        }
        preOrder(root.left);
    }

    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }

}
